### ISSUE TITLE
Allow for custom standard-engine linters not on the hard-coded list

### ISSUE THREAD

**Issue by spudly** [gustavnikolaj/linter-js-standard-engine#9]
So I have a unique situation. Where I work, we have our own code style, and I created a linter for us that uses standard-engine, but I can't open-source it because of company policy. I also created an atom plugin based on linter-js-standard that allows us to integrate the linter with atom.

Now that I've found your repository, which I believe takes the correct approach, I think I could use it instead of maintaining my own atom plugin if only my own linters name was on the supported list, but for that to work you'd either have to add my linter to your supported linters list (even though it's not open source) or you'd have to add a setting for adding additional supported standard-engine linters.

I believe the latter would be best, but am not sure if you want to maintain something like that once it's in your repo. It would also allow other people to create linters and use your plugin without needing to change your code to have support for it.

Would you accept a pull request adding a configurable setting for additional supported standard-engine linters?

**Comment by gustavnikolaj** [comment]
Hi @spudly,

This plugin is not yet completed. It started as I was incredibly frustrated with the way the other existing plugin worked. I wanted it to be faster, leaner and not constrain you to use the version of the specific linters bundled with the plugin.

So your idea is very much in line with that philosophy. I actually wanted to do something similar - I was thinking about adding a plugin config in Atom that allowed you to extend the list - would something like that work?

## 

As I mentioned this plugin is not yet mature - I kind of gave up on standard-engine, as I realized that it didn't really provide much more than what eslint gave me out of the box. Except from making any customization, besides globals, impossible. While that is a nice feature for open source projects that are new and simple, my situation is quite similar to yours. I thought I would offer my experience and perspective...

At work we have multiple web applications - some are 5+ years old, others are not yet shipped to production. Given those circumstances, I found myself with two options:
1. Creating a very permissive standard-engine based solution (that had to allow all the stuff that all of our applications did - refactoring all existing code to modern standards is not an option).
2. Opt out of standard-engine and create a set of sharable eslint configurations that can be combined in each project, to get a more strict and fitting linter.

I started with option 1, as I had used standard on a few personal projects, and really liked it. The (incomplete) result can be seen here: https://github.com/gustavnikolaj/onelint

I gave up on that. Primarily for two reasons:
- It ended up having to be way too permissive if it were to be used in all of our projects - and it should - the initial goal was to get rid of all the different permutations of linter setups we had.
- Editor plugins proved to be a PITA to put it mildly. I realized that I had to pretty much do it myself, if I wanted sane setups.

I ended up going for the second option. There's already a plugin for most editors for ESLint. Given the nature of ESLint, the configuration is read from the project, and not a potentially different version of the linter bundled with the plugin. And it allows me to break the configuration out into smaller pieces. I have mostly finished migrating our existing jscs/jshint based setup, which will be the basis shared between all projects: https://github.com/One-com/eslint-config-onelint

I will have to do a little more work documenting the setup in the project - but it will not be much more harder than standard-engine - and I think that it'll actually end up being a lot simpler.

So unfortunately, I don't currently see myself working on this plugin in the future. You're welcome to take over if you choose to stick to your standard-engine approach.

I hope that my history can be useful for you, and if you would like to discuss it further, feel free to get in touch.

**Comment by spudly** [comment]
>  I actually wanted to do something similar - I was thinking about adding a plugin config in Atom that allowed you to extend the list - would something like that work?

That's exactly what I want.

My situation is a little different. We are migrating our codebase from coffeescript to es6, so we don't have the same constraints, so standard-engine works best for us. The reason we don't use ESLint directly is that it requires significantly more setup and we don't want to be configuring different rules for each project.

**Comment by gustavnikolaj** [comment]
I'd gladly accept a PR adding support for that :-)

## 

The usage section of the readme in the shared eslint configuration that I linked to describes the setup. I don't think that it is that much more complex: https://github.com/One-com/eslint-config-onelint#usage

You could also choose to use a `eslintConfig` field in a package.json file instead, which would make the entire diff against a standard based project:

``` json
{
  "name": "somepackage",
  "dependencies": {
    "onelint": "x"
  }
}

```

vs.

``` json
{
  "name": "somepackage",
  "dependencies": {
    "eslint": "y",
    "eslint-config-onelint": "z",
  },
  "eslintConfig": {
    "extends": [ "onelint" ]
  }
}
```

**Comment by ppvg** [comment]
I actually implemented a [crude fix](https://github.com/gustavnikolaj/linter-js-standard-engine/pull/10) for this before reading the issue. Oops. :grin:

Our main reason for going with standard-engine was that our eslintrc's were getting towards 20 lines, extending about three or four configs. But I guess we could bundle that into one custom config, bringing it down to 5 lines in package.json (versus 1 with the standard-engine-based linter). That doesn't sound too bad. Something to think about...

For now, though, I'd be pretty handy if we could use this plugin without our custom engine. :smile:

**Comment by gustavnikolaj** [comment]
I have commented on your PR, and as said above, I'm more than happy to merge a PR adding support for this. The easiest fix would of course just have been adding the name of your linter to the supported linters list. But this is a less lazy change :-)

I am considering picking this plugin up again, as I'm a bit annoyed with linter-eslint (which is the atom plugin for vanilla eslint). It's a bit slower than this plugin, even when running less code in the linter it self. So I'm considering adding vanilla eslint support for this, as that would allow me to use standard and eslint based projects with the same linter. No promises though!

If I ultimately decide against it, I'll be happy to handover this plugin to someone who would want to maintain it.

**Comment by gustavnikolaj** [comment]
Fixed by the introduction of the `standard-engine` property in `package.json`. Released with version 0.1.0 of the plugin, after merging #14 / #15
